File: algorithms/insertion_sort/insertion_sort.py
def insertion_sort(arr: list[int], size: int) -> None:
    for i in range(1, size):
        key = arr[i]
        j = i - 1

        while (j >= 0 and arr[j] > key):
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


def insertion_sort_recursive(arr: list[int], size: int) -> None:
    bitSize = size - 1
    if size > 1:
        insertion_sort_recursive(arr, bitSize)
    
    key = arr[bitSize]
    j = bitSize - 1

    while (j >= 0 and arr[j] > key):
        arr[j + 1] = arr[j]
        j -= 1
    arr[j + 1] = key


def main():
    from random import randint
    arr = [randint(1, 100) for _ in range(10)]
    print(f"Original array:\n{arr}")

    insertion_sort(arr, len(arr))
    # uncomment the following line for the recursive version of this algorithm
    # insertion_sort_recursive(arr, len(arr))
    print(f"New array:\n{arr}")


if __name__ == "__main__":
    main()

File: algorithms/insertion_sort/insertion_sort.c
#include<stdio.h>
#include<stdlib.h>


void insertion_sort(int *arr, int size) {
    for (int i = 1; i < size; i++) {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}


void insertion_sort_recursive(int *arr, int size) {
    int bitSize = size - 1;
    if (size > 1) {
        insertion_sort_recursive(arr, bitSize);
    }

    int key = arr[bitSize];
    int j = bitSize - 1;

    while (j >= 0 && arr[j] > key) {
        arr[j + 1] = arr[j];
        j--;
    }
    arr[j + 1] = key;
}


int main(int argc, char *argv[]) {
    int n = argc - 1;
    int *arr = malloc(n * sizeof(int));

    for (int i = 1; i < n; i++) {
        char *endPtr;
        arr[i - 1] = strtol(argv[i], &endPtr, 10);
    }

    printf("Original array:\n");
    for (int i = 0; i < n; i++) {
        printf("%d,\t", arr[i]);
    }

    insertion_sort(arr, n);
    // Uncomment the following line for the recursive version of this algorithm
    // insertion_sort_recursive

    printf("New array:\n");
    for (int i = 0; i < n; i++) {
        printf("%d,\t", arr[i]);
    }
}
